src/System.Private.CoreLib/shared/System/DateTime.Windows.cs

```diff
@@ -0,0 +1,226 @@
+
+                int year, month, day;
+                dt.GetDatePart(out year, out month, out day);
+
+                systemTime.Year = (ushort) year;
+                systemTime.Month = (ushort) month;
+                systemTime.DayOfWeek = (ushort) dt.DayOfWeek;
+                systemTime.Day = (ushort) day;
+                systemTime.Hour = (ushort) dt.Hour;
+                systemTime.Minute = (ushort) dt.Minute;
+                systemTime.Second = (ushort) dt.Second;
+                systemTime.Milliseconds = (ushort) dt.Millisecond;
+                hundredNanoSecond = 0;
+            }
+        };
+
+#if !CORECLR
+        internal static readonly bool s_systemSupportsPreciseSystemTime = SystemSupportsPreciseSystemTime();
+
+        private static unsafe bool SystemSupportsPreciseSystemTime()
+        {
+            if (Environment.IsWindows8OrAbove)
+            {
+                // GetSystemTimePreciseAsFileTime exists and we'd like to use it.  However, on
+                // misconfigured systems, it's possible for the "precise" time to be inaccurate:
+                //     https://github.com/dotnet/coreclr/issues/14187
+                // If it's inaccurate, though, we expect it to be wildly inaccurate, so as a
+                // workaround/heuristic, we get both the "normal" and "precise" times, and as
+                // long as they're close, we use the precise one. This workaround can be removed
+                // when we better understand what's causing the drift and the issue is no longer
+                // a problem or can be better worked around on all targeted OSes.
+
+                long systemTimeResult;
+                Interop.Kernel32.GetSystemTimeAsFileTime(&systemTimeResult);
+
+                long preciseSystemTimeResult;
+                Interop.Kernel32.GetSystemTimePreciseAsFileTime(&preciseSystemTimeResult);
+
+                return Math.Abs(preciseSystemTimeResult - systemTimeResult) <= 100 * TicksPerMillisecond;
+            }
+
+            return false;
+        }
+
+        private static unsafe bool ValidateSystemTime(in Interop.Kernel32.SYSTEMTIME time, bool localTime)
+        {
+            fixed (Interop.Kernel32.SYSTEMTIME* timePtr = &time)
+            {
+                if (localTime)
+                {
+                    Interop.Kernel32.SYSTEMTIME st;
+                    return Interop.Kernel32.TzSpecificLocalTimeToSystemTime(IntPtr.Zero, timePtr, &st) != Interop.BOOL.FALSE;
+                }
+                else
+                {
+                    long timestamp;
+                    return Interop.Kernel32.SystemTimeToFileTime(timePtr, &timestamp) != Interop.BOOL.FALSE;
+                }
+            }
+        }
+
+        private static unsafe bool FileTimeToSystemTime(long fileTime, out FullSystemTime time)
+        {
+            time = new FullSystemTime();
+            fixed (Interop.Kernel32.SYSTEMTIME* systemTimePtr = &time.systemTime)
```

Comment:
All these SYSTEMTIME and similar structures are allocated on stack. None of this pinning should be needed. I would just pass the pointers from the layer above.